### PR TITLE
Split sparse index read into chunks based on optimal I/O size for the media

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/encodings/plain.rs [hunk 1]
```diff
@@ -238,13 +220,41 @@ impl<'a> Decoder for PlainDecoder<'a> {
...
-        let adjusted_offsets = subtract_scalar(indices, start)?;
-        Ok(take(&array, &adjusted_offsets, None)?)
+        let arrays = stream::iter(chunk_ranges)
+            .map(|cr| async move {
+                let index_chunk = indices.slice(cr.start as usize, cr.len());
+                let request: &UInt32Array = as_primitive_array(&index_chunk);
+
+                let start = request.value(0);
+                let end = request.value(request.len() - 1);
+                let array = self.get(start as usize..end as usize + 1).await?;
```
Comment (changhiskhan) [inline rust/src/encodings/plain.rs:249]: not sure what's Rust best practice - is it better to do ..= or .. and +1?

File: rust/src/arrow.rs [hunk 2]
```diff
@@ -98,6 +100,31 @@ impl DataTypeExt for DataType {
     fn is_dictionary(&self) -> bool {
         matches!(self, Self::Dictionary(_, _))
     }
+
+    fn byte_width(&self) -> usize {
+        match self {
+            DataType::Int8 => 1,
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/arrow.rs:106]: ByteWidth is on the `Int8/16/32Type`, yes, i guess we can , but it is more code than this. I am ok either way.

File: rust/src/encodings/plain.rs [hunk 3]
```diff
@@ -238,13 +220,41 @@ impl<'a> Decoder for PlainDecoder<'a> {
         if indices.is_empty() {
             return Ok(new_empty_array(self.data_type));
         }
+        let block_size = self.reader.prefetch_size() as u32;
+        let byte_width = self.data_type.byte_width() as u32;
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/encodings/plain.rs:224]: It should be fine if byte_width is bigger than block_size, the loop still works?

File: rust/src/encodings/plain.rs [hunk 4]
```diff
@@ -238,13 +220,41 @@ impl<'a> Decoder for PlainDecoder<'a> {
...
+
+        let mut chunk_ranges = vec![];
+        let mut start: u32 = 0;
+
+        for idx in 0..(indices.len() - 1) as u32 {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/encodings/plain.rs:229]: Yes. That's right.

File: rust/src/encodings/plain.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (changhiskhan) [reply rust/src/encodings/plain.rs:224]: Ok yeah if one row is so big then just fetch each row then. Makes sense

File: rust/src/encodings/plain.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs:249]: `..=` is actually `inclusive`, where  `a..=b` equals to `a..b-1`?

**Review by changhiskhan — APPROVED** [review]
Shipit

**Pull request merged**